src/IonDecimal.ts

```diff
@@ -138,22 +138,22 @@ export class Decimal {
      * returns a new Decimal object corresponding to the value.  If a string such as
      * '5' is provided, a 'd0' suffix is assumed.
      */
-    static parse(str: string): Decimal {
+    static parse(str: string): Decimal | null {
         let exponent = 0;
         if (str === 'null' || str === 'null.decimal') return null;
         let d = str.match('[d|D]');
         let exponentDelimiterIndex = str.length;
-        if (d) {
-            exponent = Number(str.substring(d.index + 1, str.length));
-            exponentDelimiterIndex = d.index;
+        if (d !== undefined && d !== null) {
+            exponent = Number(str.substring(d.index! + 1, str.length));
+            exponentDelimiterIndex = d.index!;
         }
         let f = str.match('\\.');
         let coefficientText: string;
         if (f) {
-            let exponentShift = d ? (d.index - 1) - f.index : (str.length - 1) - f.index;
+            let exponentShift = d ? (d.index! - 1) - f.index! : (str.length - 1) - f.index!;
```

Comment:
I wonder if `tsc` is smart enough to minimize the number of `!`s that actually expand to non-null checks or if we're paying for that branch+throw every time it appears.

Reply:
The not-null assertions appear to be compile-time only.  For example, the generated JavaScript of this line is simply:
```javascript
let exponentShift = d ? (d.index - 1) - f.index : (str.length - 1) - f.index;
```

And there's no generated logic to assert non-null-ness at runtime in this block.